coderd/autobuild/lifecycle_executor_test.go

```diff
@@ -72,6 +73,74 @@ func TestExecutorAutostartOK(t *testing.T) {
 	require.Equal(t, template.AutostartRequirement.DaysOfWeek, []string{"monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"})
 }
 
+func TestMultipleLifecycleExecutors(t *testing.T) {
+	t.Parallel()
+
+	db, ps := dbtestutil.NewDB(t)
+
+	var (
+		sched = mustSchedule(t, "CRON_TZ=UTC 0 * * * *")
+		// Create our first client
+		tickChA  = make(chan time.Time)
+		statsChA = make(chan autobuild.Stats)
+		clientA  = coderdtest.New(t, &coderdtest.Options{
+			IncludeProvisionerDaemon: true,
+			AutobuildTicker:          tickChA,
+			AutobuildStats:           statsChA,
+			Database:                 db,
+			Pubsub:                   ps,
+		})
+		// ... And then our second client
+		tickChB  = make(chan time.Time)
+		statsChB = make(chan autobuild.Stats)
+		_        = coderdtest.New(t, &coderdtest.Options{
+			IncludeProvisionerDaemon: true,
+			AutobuildTicker:          tickChB,
+			AutobuildStats:           statsChB,
+			Database:                 db,
+			Pubsub:                   ps,
+		})
+		// Now create a workspace (we can use either client, it doesn't matter)
+		workspace = mustProvisionWorkspace(t, clientA, func(cwr *codersdk.CreateWorkspaceRequest) {
+			cwr.AutostartSchedule = ptr.Ref(sched.String())
+		})
+	)
+
+	// Have the workspace stopped so we can perform an autostart
+	workspace = coderdtest.MustTransitionWorkspace(t, clientA, workspace.ID, database.WorkspaceTransitionStart, database.WorkspaceTransitionStop)
+
+	// Get both clients to perform a lifecycle execution tick
+	next := sched.Next(workspace.LatestBuild.CreatedAt)
+	go func() {
+		tickChA <- next
+		close(tickChA)
+	}()
+	go func() {
+		tickChB <- next
+		close(tickChB)
+	}()
```

Comment:
Is this potentially racy? We're testing that the lock acquire works but theoretically that might not happen if the first coderd grabs the job, completes it, and then the second one does.

I doubt it matters as I suppose we're happy even if the try acquire is hit only a faction of the time, but thought I'd flag it anyway.

Reply:
Looking again, you're probably right. I ran the test with verbose logging and it looks like this all occurs within `0.05s`.

If the test _doesn't_ hit the lock, then we are likely to hit a flake. I'll have a go at increasing this time buffer.

Reply:
I think you might be able to reduce (but not eliminate) racyness by having a second `chan struct{}` that you then close after starting both goroutines, making them both wait until it's closed to start.

e.g.

```
    startCh := make(chan struct{})
	go func() {
        <-startCh
		tickChA <- next
		close(tickChA)
	}()
	go func() {
        <-startCh
		tickChB <- next
		close(tickChB)
	}()
    close(startCh)
```

You might also be able to get both of them to tick very closely in time by sharing the same tick channel, and making it buffered with size 2.

Reply:
I've gone with your proposal @johnstcn. 

It looks like for testing we just use an echo provisioner job, so getting that to take artificially longer for this specific test may not be a trivial task.